sync controller: replace revision file with full diff each interval

skriss opened the pull request:
Signed-off-by: Steve Kriss <krisss@vmware.com>

Closes #1343 

This is kind of what I was thinking ref. https://github.com/heptio/velero/issues/1343#issuecomment-519605161.  Instead of relying on the `revision` file to tell us when BSL contents change, we get the full list of backups from the BSL (this is just a single API call to the object store), and do a setwise comparison (based on name) to the backups in the cluster. We then sync/import any ones that are in the BSL but not in the cluster.

I also removed the limitation that the sync period must be >= 1 minute - I figure users should be able to set it lower if they want.

I haven't yet removed the actual updating of the `revision` file when backups happen - we could choose to leave that in for now, or remove it entirely.

If this makes sense, I'll proceed with tests/etc.

prydonius commented on pkg/controller/backup_sync_controller.go:
```diff
@@ -162,44 +134,44 @@ func (c *backupSyncController) run() {
-			log = log.WithField("backup", backupName)
-			log.Debug("Checking this backup to see if it needs to be synced into the cluster")
+		// get a list of all the backups that exist as custom resources in the cluster
+		clusterBackups, err := c.backupLister.Backups(c.namespace).List(labels.Everything())
```
curious: what's the difference between `labels.Everything()` and an empty `metav1.ListOptions{}`?

skriss replied:
this is using a Lister, which doesn't take a `ListOptions` like the clients do -- it only takes a label selector. So same effect, but different types.

skriss commented:
I fixed the tests, and removed the code that gets/sets the `metadata/revision` file in object storage.

The only other thing to consider doing is adding some code to remove any existing `metadata/revision` files, or we could just add a script/instructions to the release notes to do this - any thoughts?

carlisia commented:
I'm fine with leaving the `revision` file in backups from past versions.

nrb commented on pkg/controller/backup_sync_controller.go:
```diff
@@ -66,10 +65,10 @@ func NewBackupSyncController(
 ) Interface {
-	if syncPeriod < time.Minute {
-		logger.Infof("Provided backup sync period %v is too short. Setting to 1 minute", syncPeriod)
+	if syncPeriod <= 0 {
```
Doesn't this conditional keep the effective minimum sync time at 1 minute?

skriss replied:
No - you can explicitly specify a sync period of e.g. `15s`. This just defaults it to 1 minute if a 0/negative value was provided.

skriss commented:
> Have you observed any time differences (based on log messages would be simplest, I suppose) between the old approach and this one?

Haven't specifically looked, though performance seemed fine/~instantaneous in the testing I did. I can try with some bigger data sets.

skriss commented:
Tweaked logging so it only info-logs if there are actually backups to sync into the cluster for a location; otherwise, debug-logs.

skriss commented:
> Have you observed any time differences (based on log messages would be simplest, I suppose) between the old approach and this one?

The algorithm I'm using now should actually be faster than the old one -- previously, when syncing a location, we'd first get a list of all backups in the bucket, then make a K8s API call *per backup in the BSL* to check if that backup exists in-cluster.  Now, I'm making a single **List** call (to a lister) upfront to get all backups in the cluster, and then doing an in-memory set diff to figure out which ones should be synced in. So a bunch of over-the-wire API calls are saved.

carlisia commented:
@nrb PTAL.
